hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/lib/MultipleOutputs.java

```
@@ -527,9 +558,41 @@ public void collect(Object key, Object value) throws IOException {
    * @throws java.io.IOException thrown if any of the MultipleOutput files
    *                             could not be closed properly.
    */
-  public void close() throws IOException {
+  public void close() throws IOException, InterruptedException {
+    int nThreads = conf.getInt(MRConfig.MULTIPLE_OUTPUTS_CLOSE_THREAD_COUNT,
+        MRConfig.DEFAULT_MULTIPLE_OUTPUTS_CLOSE_THREAD_COUNT);
+    AtomicBoolean encounteredException = new AtomicBoolean(false);
+    ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("MultipleOutputs-close")
+        .setUncaughtExceptionHandler(
```

`IOException` is now being propagated back to the caller of `close()`, but any unexpected (unchecked) exceptions would be swallowed in this uncaught exception handler. This is different from the existing code, where the caller would receive the unchecked exception.

I think the best we can do here is to set `encounteredException` from within the uncaught exception handler, resulting in throwing the `IOException` at the bottom of the method. The message would need to be generalized to "encountered exception during close", not mentioning `IOException`, because it might also have been some other unchecked exception.

---

I understand your concern. I hope below makes sense.

```
    ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("MultipleOutputs-close")
        .setUncaughtExceptionHandler(((t, e) -> {
          LOG.error("Thread " + t + " failed unexpectedly", e);
          encounteredException.set(true);
        })).build();
```

---

Yes, makes sense. Thanks!

---

Thanks. Addressed this in my next commits.